modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java

```diff
@@ -480,6 +550,183 @@ public int getZoneId(String name) {
+     * @param zoneId Zone id.
+     * @param topVer Topology version.
+     * @return The data nodes future.
+     */
+    public CompletableFuture<Set<String>> getDataNodes(int zoneId, long topVer) {
+        CompletableFuture<Set<String>> dataNodesFut = new CompletableFuture<>();
+
+        CompletableFuture<Void> topVerFut;
+
+        synchronized (dataNodesMutex) {
+            dataNodes.putIfAbsent(zoneId, new DataNodes());
+
+            if (topVer > lastTopVer) {
+                topVerFut = topVerFutures.computeIfAbsent(topVer, key -> new CompletableFuture<>());
+            } else {
+                topVerFut = completedFuture(null);
+            }
+        }
+
+        boolean immediateScaleUp0 = false;
+        boolean immediateScaleDown0 = false;
+
+        if (zoneId == DEFAULT_ZONE_ID) {
+            immediateScaleUp0 = zonesConfiguration.defaultDistributionZone().dataNodesAutoAdjustScaleUp().value() == 0;
+            immediateScaleDown0 = zonesConfiguration.defaultDistributionZone().dataNodesAutoAdjustScaleDown().value() == 0;
+        } else {
+            NamedConfigurationTree<DistributionZoneConfiguration, DistributionZoneView, DistributionZoneChange> zones =
+                    zonesConfiguration.distributionZones();
+
+            for (int i = 0; i < zones.value().size(); i++) {
+                DistributionZoneView zone = zones.value().get(i);
+
+                if (zone.zoneId() == zoneId) {
+                    immediateScaleUp0 = zone.dataNodesAutoAdjustScaleUp() == 0;
+                    immediateScaleDown0 = zone.dataNodesAutoAdjustScaleDown() == 0;
+
+                    break;
+                }
+            }
+        }
+
+        boolean immediateScaleUp = immediateScaleUp0;
+        boolean immediateScaleDown = immediateScaleDown0;
+
+        if (!immediateScaleUp && !immediateScaleDown) {
+            synchronized (dataNodesMutex) {
+
+                DataNodes dataNodes = this.dataNodes.get(zoneId);
+
+                Set<String> nodes;
+
+                if (dataNodes == null) {
+                    nodes = emptySet();
+                } else {
+                    nodes = this.dataNodes.get(zoneId).nodes();
+                }
+
+
+                return completedFuture(nodes);
+            }
+        }
+
+        CompletableFuture<Void> topVerScaleUpFut = new CompletableFuture<>();
+
+        topVerFut.thenAccept(ignored -> {
+            synchronized (dataNodesMutex) {
+                CompletableFuture<Void> topVerScaleUpFut0 = null;
+
+                if (immediateScaleUp) {
+                    Map.Entry<Long, Long> scaleUpRevisionEntry = topVerAndScaleUpRevision.ceilingEntry(topVer);
+
+                    Long scaleUpRevision = null;
+
+                    if (scaleUpRevisionEntry != null) {
+                        scaleUpRevision = scaleUpRevisionEntry.getValue();
+                    }
+
+                    if (scaleUpRevision != null && dataNodes.get(zoneId).scaleUpRevision() < scaleUpRevision) {
```

Comment:
seems like it is possible that while we were waiting for completion of `topVerFut`, zone was deleted, potential NPE here

Reply:
Thanks. Fixed it here and other places.